### ISSUE TITLE
Migrate from MySQL to Postgres

### ISSUE THREAD

**Issue by a97marbr** [HGustavs/LenaSYS#1869]
We want to migrate from MySQL to Postgres for a number of reasons:
- Improved stability
- Improved performance
- Native support for JSON

**Comment by a16jonis** [comment]
@a97marbr @HGustavs Is this issue still actual?

**Comment by a17robhe** [comment]
See https://github.com/HGustavs/LenaSYS/issues/1871

**Comment by a97marbr** [comment]
I agree that migrating is a tedious task, BUT, it won't be done if it is never started.

Perhaps split this into multiple sub-issues? Start by identify which MySQL specific database operations we currently use ... and try to remove those?

**Comment by a18kaswi** [comment]
`sudo grep -rnw '/var/www/group4/G4-2020-W21' -e .*mysql.* > grep_mysql.txt
`http://group4.webug.his.se:20001/grep_mysql.txt

`sudo find . -type f -name "*.sql" > sql_files.txt`
http://group4.webug.his.se:20001/sql_files.txt

Here is the files referencing mysql and sql-files currently in the weekly branch.

**Comment by a18kaswi** [comment]
Tried messing around a little bit with pgloader (https://pgloader.readthedocs.io/en/latest/) didnt go to well.

**Comment by a18kaswi** [comment]
Line 43 in /shared/database.php is the function pdoConnect() this is used to connect to the database. This connection(socket?) is then used throughout the system.

http://group4.webug.his.se:20001/pdoConnect.txt 

We probably have to check every function using this connection to see if it contains mysql-specific syntax. Other than that it should not be a problem to rewrite pdoConnect() to instead of connecting to mysql connect to postgres.

**Comment by a18kaswi** [comment]
Right now the pdoConnect() function creates the socket(connection) and set it as the $pdo variable. This variable are as seen below used all over the place (lot of possible mysql-specific syntax).

http://group4.webug.his.se:20001/grep_pdo-variabel.txt

**Comment by a18kaswi** [comment]
Divided the big grep_pdo to smaller chunks. http://group4.webug.his.se:20001/SQL-migration/

**Comment by a18kaswi** [comment]
Finally these are the rows creating a new connection to the database (new PDO)

http://group4.webug.his.se:20001/SQL-migration/grep_new-pdo.txt

**Comment by a18kaswi** [comment]
First connection between LenaSYS and database is in the installer. The socket $connection is used to run queries from the files inside LenaSYS\install\SQL.

So first of all we need to translate the installer and corresponding files to postgres-syntax. This is two tasks: 

Rewrite install.php, for example there is no `FLUSH PRIVILEGES` or `IDENTIFIED BY` syntax in Postgres, probably more of these cases.

And translate files inside LenaSYS\install\SQL, for example 
> MySQL uses ' or " to quote values (i.e. WHERE name = "John"). This is not the ANSI standard for databases. PostgreSQL uses only single quotes for this (i.e. WHERE name = 'John').
--
https://en.wikibooks.org/wiki/Converting_MySQL_to_PostgreSQL

This takes care of how the test-data, i.e all data with a fresh install, gets into the (currently) MySQL-database.

**Comment by a18kaswi** [comment]
Then we need to take care about how new data are created and added into the database. 

The heavy lifting is as mentioned before done by the socket $pdo thats created by pdoConnect() in /Shared/database.php

First of all we have to rewrite pdoConnect() so that $pdo connects to PostgreSQL instead of MySQL. 

Further we have to make sure that all subsequent use of $pdo conforms to PostgreSQL-syntax.

Besides the $pdo socket there are some minor ones in use: http://group4.webug.his.se:20001/SQL-migration/grep_new-pdo.txt

**Comment by a18kaswi** [comment]
TLDR: 

Rewrite creation of database-connections (sockets) and make sure functions using these conform to PostgreSQL-syntax. 

Make sure files imported by installer conform to PostgreSQL-syntax. 

Am I missing something here @a97marbr @HGustavs

**Comment by a18kaswi** [comment]
Have split this issue into:

#9502 
#9503 
#9504 
#9505 
#9506

**Comment by a18kaswi** [comment]
Remember to uncomment 
extension=pdo_pgsql
extension=pgsql

in php.ini !! @a18jonkl @a17robhe

**Comment by a18kaswi** [comment]
If we succeed: maybe https://bitnami.com/stack/wapp for next years students.

**Comment by a17robhe** [comment]
The init_db file is now working in postgres.

Next up is the test data:

https://github.com/HGustavs/LenaSYS/issues/9580
https://github.com/HGustavs/LenaSYS/issues/9581
https://github.com/HGustavs/LenaSYS/issues/9582
https://github.com/HGustavs/LenaSYS/issues/9583
https://github.com/HGustavs/LenaSYS/issues/9584
https://github.com/HGustavs/LenaSYS/issues/9585
https://github.com/HGustavs/LenaSYS/issues/9586
https://github.com/HGustavs/LenaSYS/issues/9588

**Comment by a97marbr** [comment]
> TLDR:
> 
> Rewrite creation of database-connections (sockets) and make sure functions using these conform to PostgreSQL-syntax.
> 
> Make sure files imported by installer conform to PostgreSQL-syntax.
> 
> Am I missing something here @a97marbr @HGustavs (of the top of your head)?

Not that I can think of

**Comment by a97marbr** [comment]
I guess that making the SQL in LenaSYS DB agnostic is the first important step. 

Do you think that, once fully converted to none vendor specific SQL, that a transition from MySQL to Postgres is possible for a live system? Or would only new installations be able to use Postgres?

**Comment by a18kaswi** [comment]
> I guess that making the SQL in LenaSYS DB agnostic is the first important step.
> 
> Do you think that, once fully converted to none vendor specific SQL, that a transition from MySQL to Postgres is possible for a live system? Or would only new installations be able to use Postgres?

Hello @a97marbr ,

### Our guess is that what you mean with the question "Do you think that, once fully converted to none vendor specific SQL, that a transition from MySQL to Postgres is possible for a live system?" is that you can migrate the current MySQL-database holding all students, graded duggas etc, into the newly reated PostgreSQL-database.

Our answer to that is yes is seems possible, but we have to go with the unadvised usage of enclosing reserved keywords in brackets, to be able to do this automatically.

Currently the names 'user' and 'groups' are used for tables in the MySQL-database. In PostgreSQL these are reserved keywords and can not be used intuitively. This is the major problem as we see it for a safely migration on the live system. 

In postgres one can circumvent this by enclosing the words in bracket i.e "groups" when creating the relation and later accessing it. From what we've gathered this is not recommended. So for now we are renaming 'user' to 'lenasys_user' and 'groups' to 'lenasys_groups' for testing purposes.

BUT we tinkered around a little. We created a postgres-database on our testserver named 'test'. We then ran our postgres-translated init.db-file against this database to create all relations. Including a relation named "user". 

![before pgloader](https://user-images.githubusercontent.com/62878146/82803940-aede5e80-9e81-11ea-9ac7-5ffc3ddcfaa1.png)

We now have a fresh database-install with the three admins in the "user" relations. Note here that we had to use the row 

`INSERT INTO "user"(username, password, newpassword, creator, superuser) values ('Grimling','$2y$12$stG4CWU//NCdnbAQi.KTHO2V0UVDVi89Lx5ShDvIh/d8.J4vO8o8m', 0, 1, 1);` 

Referencing "user" not user when adding a user. 

Next we run the pgloader.load-file which migrates an MySQL-database to an PostgreSQL-database.

`LOAD DATABASE
        FROM mysql://root:galvaniseradapa@localhost/group4test
        INTO postgresql://postgres:galvaniseradapa@localhost/test

  alter schema 'group4test' rename to 'public';
`

Ending up with the result:

![after pgloader](https://user-images.githubusercontent.com/62878146/82804127-067cca00-9e82-11ea-9b0e-c83403b2ad6d.png)

**Comment by a18kaswi** [comment]
**Just posting some comments here that can be useful in the future. These were pinned in an WIP channel so probably a little unclear but time is running out.**

Removed --->    CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;

Create table user_group uses ‘ENGINE=InnoDB DEFAULT CHARSET=latin1;’

Komment – [text] */

AUTO_INCREMENT                    SERIAL
DATETIME                 TIMESTAMP WITHOUT TIME ZONE
TINYINT(x)                SMALLINT
decimal(x,y)                NUMERIC(x, y)
BOOLEAN 1                POSTSQL cant cast integers use true/false


Discuss:

Create table user_group uses ‘ENGINE=InnoDB DEFAULT CHARSET=latin1;’

Row 45: Updated     TIMESTAMP WITHOUT TIME ZONE DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
https://stackoverflow.com/questions/1035980/update-timestamp-when-row-is-updated-in-postgresql

Line 79: ‘groups’ varchar(256). Removed the quotation marks

Line 229: templateid    INTEGER  NOT NULL,.  Changed to :emplateid    INTEGER  NOT NULL UNIQUE

Line 445-447: Removed ‘KEY’ as it seems to be legacy compadability function. 

Line 372: changed motd to ‘text’  from varchar(4086)

After changing back to user syntax got extra sad.

In a php file user must be in "user" and for that to happen the string must be in 'SELECET * FROM "user" '. if something is quoted inside the string it must be changed from '' to "" also.
if user is not in "" it will call the ?? default?? database postgres and not find the column you are looking for.

When you are looking for a specific value in user as `user.cid` the correct syntax is 'SELECT "user".uid FROM "user"'

**Comment by a18kaswi** [comment]
Closing this issue and starting a new with current situation, referencing this one.
